rfc/operation-cache-warmer.md

```diff
@@ -0,0 +1,107 @@
+---
+title: "Operation Cache Warmer"
+author: Dustin Deus
+date: 2024-08-25
+status: Draft
+---
+
+# Operation Cache Warmer
+
+- **Author:** Dustin Deus
+- **Date:** 2024-08-25
+- **Status:** Draft
+
+## Abstract
+
+This RFC describes a new feature to reduce the latency of the system by pre-planning the most expensive and requested operations before the router accepts traffic. We achieve this by computing the Top-N GraphQL operations available and making them available to all routers instances before they accept traffic.
+
+## Motivation
+
+GraphQL is a powerful tool to query data from a server. However, the flexibility of the query language comes with a cost. The cost is the complexity of the query and how expensive it is to normalize, plan and execute it. While execution performance is primarily a concern of the underlying subgraphs, the planning phase can be a unpredictable and significant latency contributor. The operation cache warmer aims to reduce this latency by pre-planning the most expensive and requested operations ahead to make it invisible to the user.
+
+# Proposal
+
+The distributed operation cache is semi-automatic and allows the user to push specific operations to the cache but also automatically computes the most expensive and requested operations of the last time frame (configurable). The cache has a fixed size of operations e.g. 100 (configurable) and is shared across all router instances. An operation can be a regular query, subscription, mutation or persisted operation. When the cache capacity is reached, manual operations have a higher priority than automatic operations. This allows users to manage the priority of operations in the cache themselves. It is possible that operations aren't compatible with all future schema changes. In that case, the operation is removed from the cache.
+
+### Pushing operations to the cache
+
+The User can push individual operations to the operation cache by using the CLI:
+
+```bash
+wgc federated-graph operation-cache add --graph mygraph --file operations.json
```

Comment:
Curious about manual cache invalidation.

What will happen in the following scenario:

First, we run:

```
wgc federated-graph operation-cache add --graph mygraph --file operationA.json
```

Then we run:
```
wgc federated-graph operation-cache add --graph mygraph --file operationB.json
```

What will we have in the cache after the second operation? Will it be just Operation B or both Operation A and Operation B? 

Curious, because if it is both Operation A and Operation B, the cache eventually will be filled with the manual operation and there will not be space for automatic operations unless we can invalidate it explicitly.

Reply:
This is right. Both operations will be in the "batch". This is the part where a customer can manage the cache themself and is responsible of it. At some point, we could also introduce a percentage limit about how much space can be reserved for manual pushed operations. In the first version, we want to keep it simple.